### ISSUE TITLE
PPO critic never learns: `non_blocking` copy races with `add_` when building value targets

### ISSUE THREAD

**Issue by CptK** [JPK314/rlgym-learn-algos#30]
## Summary

In `PPOLearner.learn`, the value targets are built by copying into two preallocated buffers with `non_blocking=True` and then immediately reading one of them. The `add_` reads `batch_advantages` before the copy has landed, so `batch_target_values` keeps holding the raw value predictions instead of `values + advantages`.

The critic is then scored against its own output. The MSE is ~0, its gradient is ~0, so the critic never updates — which means on the next iteration its predictions again equal the targets. The result is a self-reinforcing trap: the run trains at full speed and learns essentially nothing.

This is silent. There is no error, throughput looks great, checkpoints save normally. The only signals are `Critic Loss: 0.0`, `Critic Update Magnitude: 0.0`, and actor entropy pinned at `ln(n_actions)`.

## Environment
| | |
|---|---|
| rlgym-learn | 2.0.0 |
| rlgym-learn-algos | 0.4.0 |
| rlgym | 2.0.1 |
| rocketsim | 2.2.1 |
| torch | 2.13.0 |
| numpy | 1.26.4 |
| Python | 3.12.9 |
| OS | macOS 26.6.2, arm64 (Apple M3 Pro) |
| device | `cpu` (both learner and experience buffer) |

Setup is otherwise the stock `quick_start_guide.py` wiring: `LookupTableAction` (90 discrete actions) with `RepeatAction(repeats=8)`, `DiscreteFF` + `BasicCritic` at `(256, 256, 256)`, `NumpyExperienceBuffer` with `GAETrajectoryProcessor`, `batch_size=25000`, `n_minibatches=1`, `n_epochs=1`.

## Symptoms

Every iteration, for the entire run:

```
Actor Entropy: 4.4989...
Critic Loss: 0.0
Critic Update Magnitude: 0.0
```

`4.4989 ≈ ln(90) = 4.4998`, i.e. the policy stays at maximum entropy — the entropy bonus is the only gradient reaching the actor. Advantages are zero in the actor loss too (`-min(ratio * adv, clipped * adv)` with `adv` contributing nothing meaningful), so the policy never improves.

Observed across 300k timesteps, with and without `standardize_rewards`, at several `n_proc` values.

## Offending code

`rlgym_learn_algos/ppo/ppo_learner.py`, lines 295–298:

```python
_ = self.batch_old_probs.copy_(_batch_old_probs, non_blocking=True)
_ = self.batch_target_values.copy_(_batch_values, non_blocking=True)
_ = self.batch_advantages.copy_(_batch_advantages, non_blocking=True)
_ = self.batch_target_values.add_(self.batch_advantages)   # <-- reads before the copy lands
```

## Diagnosis

Instrumenting `torch.nn.MSELoss.forward` to print the critic's inputs:

```python
_orig = torch.nn.MSELoss.forward
def probe(self, inp, tgt):
    out = _orig(self, inp, tgt)
    print(f"[critic-mse] loss={out.item():.8g} "
          f"in.absmax={float(inp.abs().max()):.6g} "
          f"tgt.absmax={float(tgt.abs().max()):.6g} "
          f"grad_fn={inp.grad_fn is not None}", flush=True)
    return out
torch.nn.MSELoss.forward = probe
```

## Suggested fix

Drop `non_blocking=True` on lines 295–297, or synchronise before line 298. These are CPU-to-CPU copies in this configuration, where `non_blocking` is documented as a no-op, so removing it should cost nothing.

## Workaround

Forcing `Tensor.copy_` synchronous before constructing the learner:

```python
import torch
_copy = torch.Tensor.copy_
def _sync_copy(self, src, non_blocking=False):
    return _copy(self, src, False)
torch.Tensor.copy_ = _sync_copy


## Caveat

I have not established *why* `non_blocking=True` misbehaves here. It is
documented as a no-op for CPU-to-CPU copies, so this may be specific to torch
2.13.0 on macOS/arm64, or to the interaction with the buffers allocated by
`torch.empty`. What is reproducible is that the copies do not land before the
`add_`, and that synchronising fixes it.

**Comment by JPK314** [comment]
Thanks for the report, I'm not sure why this happens either but if the learner device is CPU I'm just going to do non_blocking=False everywhere I previously had it as true.

**Comment by JPK314** [comment]
Fixed in 0.4.1
